### PR TITLE
[Table] scrollToRegion instance method

### LINKED ISSUE #1065: Programatically select *and* jump to row in table
Hi there,

First off, thank you for your wonderful set of components!

I was wondering if it was possible to programatically jump to a row in the data table. 

For example, say I have this function to update my state (and thus update the selected region of the table)
```js
const select10thRow = () => {
      this.setState({selectedRegions: [{rows: [10, 10]}]})
    }
```

But if the 10th row is no longer within view (because the user's scrolled), the row will be selected, but the table will not jump to it. Is there a way to get it to jump as well?

I saw this issue, but it looks like it is more for jumping on keyboard interactions:
https://github.com/palantir/blueprint/issues/915

Thanks!

### LINKED ISSUE #1430: Expose ability for app to scroll to specific row/column in table
Example usecase: App adds a search bar to search for a column, then scrolls user to that column

### PR DESCRIPTION
**Pull request by cmslewis**
#### Checklist

- [x] Include tests
- [x] Update documentation

#### Changes proposed in this pull request:

- 💎 __NEW__ `Table` now offers a __`scrollToRegion(region: IRegion)`__ instance method to programmatically scroll a desired row, column, or cell into view!
    - The target region will be positioned in the top/left corner of the viewport.
        - Or if the table is maximally scrolled, then having the target cell simply in view is enough.
    - If there are active frozen rows and/or columns:
        - Scrolling to a frozen row will move the table back to `scrollTop=0`.
        - Scrolling to a frozen column will move the table back to `scrollLeft=0`.
        - Scrolling to a frozen `TOP_LEFT` cell will move the table back to (0, 0).
        - Scrolling to a non-frozen row/column will scroll the target region to just _beyond_ the frozen regions, keeping it clearly in view.
- 🖋 Added documentation for both `Table` instance methods:
    - `resizeRowsByTallestCell`
    - `scrollToRegion`
- Also added some new `scrollTo` controls on the `Table` example page.

#### Followup work (for later)

- #1501 Add an `animated: boolean` option to animate the scroll to the new region. // CC @llorca 

#### Reviewers should focus on:

- How does this API look? Tried to be as simple and general as possible. Ended up not including an `animated` option for now, because that'd actually be a lot more work.

#### Screenshot

image

### CODE REVIEW

File: packages/table/preview/index.tsx [hunk 1]
```diff
@@ -417,7 +443,34 @@ class MutableTable extends React.Component<{}, IMutableTableState> {
             this.toTruncatedPopoverModeLabel,
             this.handleNumberStateChange,
             "enableCellTruncation",
-            true,
+            [true],
+        );
+
+        // Table: "Scroll to"
+        const scrollToRegionTypeSelectMenu = this.renderSelectMenu(
```
Comment (cmslewis) [inline packages/table/preview/index.tsx]: This whole area is a mess now. Wanted to get a preview up quickly, but would like to clean this up before merging.

File: packages/table/preview/index.tsx [hunk 2]
```diff
@@ -479,13 +541,28 @@ class MutableTable extends React.Component<{}, IMutableTableState> {
...
+                return "";
+        }
+    }
+
     private renderSwitch(
         label: string,
         stateKey: keyof IMutableTableState,
         prereqStateKey?: keyof IMutableTableState,
-        prereqStateKeyValue?: any,
+        prereqStateKeyValues?: any[],
```
Comment (cmslewis) [inline packages/table/preview/index.tsx]: Can undo all these changes; ended up going in a different direction.

File: packages/table/src/common/index.ts [hunk 3]
```diff
@@ -10,4 +10,7 @@ export { Grid } from "./grid";
 export { Rect, AnyRect } from "./rect";
 export { RoundSize } from "./roundSize";
 export { Utils } from "./utils";
-// NOTE: Errors is not exported in public API
+
+// NOTE: The following are not exported in the public API:
+// - Errors
+// - internal/
```
Comment (cmslewis) [inline packages/table/src/common/index.ts]: Thoughts on this `common/internal/` folder to clearly denote things that won't be exported?

File: packages/table/src/common/internal/scrollUtils.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
+/**
```
Comment (cmslewis) [inline packages/table/src/common/internal/scrollUtils.ts]: Really want to get into the practice of taking utility code out of `table.tsx`. I'd also like to avoid putting all utility code into one huge `Utils` file if possible. Thus, here's a new `scrollUtils` file.

File: packages/table/src/common/internal/scrollUtils.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+export function getScrollPositionForRegion(
+    region: IRegion,
+    currScrollLeft: number,
+    currScrollTop: number,
+    getLeftOffset: (columnIndex: number) => number,
+    getTopOffset: (rowIndex: number) => number,
+    numFrozenRows: number = 0,
+    numFrozenColumns: number = 0,
+) {
```
Comment (cmslewis) [inline packages/table/src/common/internal/scrollUtils.ts]: A lot of params, but this way I don't have to pass a bloated `viewportRect` or `grid` object from `<Table>`. Thoughts?

File: packages/table/src/quadrants/tableQuadrantStack.tsx [hunk 6]
```diff
@@ -192,6 +192,18 @@ export class TableQuadrantStack extends AbstractComponent<ITableQuadrantStackPro
...
     }
 
+    /**
+     * Scroll the main quadrant to the specified scroll offset, keeping all other quadrants in sync.
+     */
+    public scrollToPosition(scrollLeft: number, scrollTop: number) {
+        const { scrollContainer } = this.quadrantRefs[QuadrantType.MAIN];
+        this.wasMainQuadrantScrollChangedFromOtherOnWheelCallback = false;
+        // this will trigger the main quadrant's scroll callback below
+
```
Comment (cmslewis) [inline packages/table/src/quadrants/tableQuadrantStack.tsx]: Can delete this new line.

File: packages/table/src/quadrants/tableQuadrantStack.tsx [hunk 7]
```diff
@@ -368,12 +380,17 @@ export class TableQuadrantStack extends AbstractComponent<ITableQuadrantStackPro
     // use the more generic "scroll" event for the main quadrant, which captures both click+dragging
     // on the scrollbar and trackpad/mousewheel gestures
     private handleMainQuadrantScroll = (event: React.UIEvent<HTMLElement>) => {
+        console.log("[TableQuadrantStack] handleMainQuadrantScroll");
```
Comment (cmslewis) [inline packages/table/src/quadrants/tableQuadrantStack.tsx]: Can delete these `console.log`s.

File: packages/table/src/table.tsx [hunk 8]
```diff
@@ -1030,48 +1081,31 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
         const rowIndexEnd = showFrozenRowsOnly ? numFrozenRows : rowIndices.rowIndexEnd;
 
         return (
-            // <div
```
Comment (cmslewis) [inline packages/table/src/table.tsx]: Went ahead and deleted this unrelated commented code. Was a relic from the frozen rows/columns refactor.

File: packages/table/src/table.tsx [hunk 9]
```diff
@@ -1635,12 +1669,16 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
         const didScrollLeftChange = nextScrollLeft !== viewportRect.left;
 
         if (didScrollTopChange || didScrollLeftChange) {
-            // we need to modify the body element explicitly for the viewport to shift
+            // we need to modify the scroll container explicitly for the viewport to shift. in so
+            // doing, we add the size of the header elements, which are not technically part of the
+            // "grid" concept (the grid only consists of body cells at present).
```
Comment (cmslewis) [inline packages/table/src/table.tsx]: This was actually a regression that happened during the frozen rows/columns refactor. But all that broke was auto-scrolling as you use arrow keys to move the focus cell around. Still, works again now. 👍

File: packages/table/src/quadrants/tableQuadrantStack.tsx (hunk 7 above)
Comment (cmslewis) [inline packages/table/src/quadrants/tableQuadrantStack.tsx]: Done

File: packages/table/src/quadrants/tableQuadrantStack.tsx (hunk 6 above)
Comment (cmslewis) [inline packages/table/src/quadrants/tableQuadrantStack.tsx]: Done

File: packages/table/preview/index.tsx (hunk 2 above)
Comment (cmslewis) [inline packages/table/preview/index.tsx]: Done

File: packages/table/src/common/index.ts (hunk 3 above)
Comment (giladgray) [inline packages/table/src/common/index.ts]: 👍 yes nice

File: packages/table/src/common/internal/scrollUtils.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ * Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ */
+
+import { IRegion, RegionCardinality, Regions } from "../../regions";
+
+// Public
+// ======
```
Comment (giladgray) [inline packages/table/src/common/internal/scrollUtils.ts]: not sure these headings are useful... the keyword `export` gets the point across.

File: packages/table/src/common/internal/scrollUtils.ts (hunk 5 above)
Comment (giladgray) [inline packages/table/src/common/internal/scrollUtils.ts]: seems legit

File: packages/table/src/common/internal/scrollUtils.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        nextLeft = getAdjustedScrollPosition(leftOffset, frozenColumnsCumulativeWidth);
+    } else if (cardinality === RegionCardinality.FULL_ROWS) {
+        // scroll to the top of the row block
+        const topOffset = getTopOffset(region.rows[0]);
+        nextTop = getAdjustedScrollPosition(topOffset, frozenRowsCumulativeHeight);
+    } else if (cardinality === RegionCardinality.FULL_COLUMNS) {
+        // scroll to the left side of the column block
+        const leftOffset = getLeftOffset(region.cols[0]);
+        nextLeft = getAdjustedScrollPosition(leftOffset, frozenColumnsCumulativeWidth);
+    } else {
```
Comment (giladgray) [inline packages/table/src/common/internal/scrollUtils.ts]: this might make more sense as a `switch`? cuz that's exactly what it is, on enum values.

File: packages/table/src/common/internal/scrollUtils.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+// Private
+// =======
+
+/**
+ * Adjust the scroll position to align content just beyond the frozen region, if necessary.
+ */
+function getAdjustedScrollPosition(scrollOffset: number, frozenRegionCumulativeSize: number) {
+    return scrollOffset < frozenRegionCumulativeSize
+        ? 0
+        : scrollOffset - frozenRegionCumulativeSize;
```
Comment (giladgray) [inline packages/table/src/common/internal/scrollUtils.ts]: equivalent to `Math.min(scrollOffset - frozenRegionCumulativeSize, 0)`, i think.

perhaps, instead of this function, you can just do the subtraction inline and the `Math.min` once when building the `return` object?

File: packages/table/src/quadrants/tableQuadrantStack.tsx [hunk 13]
```diff
@@ -368,12 +380,17 @@ export class TableQuadrantStack extends AbstractComponent<ITableQuadrantStackPro
...
         if (this.wasMainQuadrantScrollChangedFromOtherOnWheelCallback) {
+            console.log("  this.wasMainQuadrantScrollChangedFromOtherOnWheelCallback == true.");
+            console.log("  Setting this.wasMainQuadrantScrollChangedFromOtherOnWheelCallback = false. RETURNING");
```
Comment (giladgray) [inline packages/table/src/quadrants/tableQuadrantStack.tsx]: 😱 🌳 logs 🌳 😱

File: packages/table/src/table.tsx [hunk 14]
```diff
@@ -590,6 +595,52 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
+     *   Scroll the top-left cell in the target region to the top-left corner of the viewport.
+     *
+     *   FULL_ROWS:
+     *   Scroll the top-most row in the target region to the top of the viewport.
+     *
+     *   FULL_COLUMNS:
+     *   Scroll the left-most column in the target region to the left side of the viewport.
+     *
+     *   FULL_TABLE:
+     *   Scroll the top-left cell to the top-left corner of the viewport.
```
Comment (giladgray) [inline packages/table/src/table.tsx]: can cut lines in a third by formatting as list.

```
- `CELLS`: words words
- `FULL_ROWS`: more words
```

Comment (cmslewis) [reply packages/table/src/table.tsx]: One of them was too long to fit and line-wrapped; this format takes more lines but keeps everything looking perfectly consistent. I'll see if I can whittle the words down and go back to a list.

File: packages/table/src/common/internal/scrollUtils.ts (hunk 12 above)
Comment (cmslewis) [reply packages/table/src/common/internal/scrollUtils.ts]: Yeah. It's a little clearer to me as is, but I can change it and just add a comment about the intent like `// if the new scroll offset falls within the frozen region, clamp it to 0`

File: packages/table/src/common/internal/scrollUtils.ts (hunk 11 above)
Comment (cmslewis) [reply packages/table/src/common/internal/scrollUtils.ts]: I like to avoid switches unless I can `return` immediately within each case; otherwise I have an extra `break;` line at the end of every case, which I find bloat-y and unsightly.

Comment (cmslewis) [reply packages/table/src/common/internal/scrollUtils.ts]: Also, TIL if you redeclare variables with the same name within different `switch` cases (e.g. `topOffset` and `leftOffset`), you get this compilation error: `[ts] Cannot redeclare block-scoped variable 'leftOffset'.`

**Comment by tgreenwatts** [comment]
seeing 1 small possible
-focus on c1 and then scroll to G, I lose my focus cell. It's selected but not focused?

**Comment by cmslewis** [comment]
@tgreenwatts I can't repro that. When I focus on C1 and then `scrollTo` column G, I still see the focus cell in C1 when I manually scroll back. What are the exact repro steps for the weirdness you're seeing?

File: packages/table/src/common/internal/scrollUtils.ts (hunk 10 above)
Comment (cmslewis) [reply packages/table/src/common/internal/scrollUtils.ts]: Removed

File: packages/table/src/common/internal/scrollUtils.ts (hunk 12 above)
Comment (cmslewis) [reply packages/table/src/common/internal/scrollUtils.ts]: Fixed

File: packages/table/preview/index.tsx [hunk 15]
```diff
@@ -683,6 +767,30 @@ class MutableTable extends React.Component<{}, IMutableTableState> {
...
+                region = Regions.column(scrollToColumnIndex);
+                break;
+            case RegionCardinality.FULL_TABLE:
+                region = Regions.table();
+                break;
+            default:
+                return;
+        }
+
+        this.tableInstance.scrollToRegion(region);
```
Comment (themadcreator) [inline packages/table/preview/index.tsx]: is invoking methods on the table ref the standard way for actions like this? do we have other components/features that do this? is this what users will expect?

File: packages/table/preview/index.tsx [hunk 16]
```diff
@@ -724,6 +832,10 @@ class MutableTable extends React.Component<{}, IMutableTableState> {
         return handleNumberChange((value) => this.setState({ [stateKey]: value }));
     }
 
+    private handleRegionCardinalityChange = (stateKey: keyof IMutableTableState) => {
+        return handleNumberChange((value) => this.setState({ [stateKey]: value }));
```
Comment (themadcreator) [inline packages/table/preview/index.tsx]: man this pattern is so common, i wish react had a shorthand for this.

Comment (themadcreator) [reply packages/table/preview/index.tsx]: e.g. 4 lines above

File: packages/table/src/common/internal/scrollUtils.ts (hunk 4 above)
Comment (themadcreator) [reply packages/table/src/common/internal/scrollUtils.ts]: that's cool, if you add more do you think it's useful to keep track of an `index.ts` that re-exports? it's a tradeoff between maintenance of index.ts and simplifying import lines.

File: packages/table/src/common/internal/scrollUtils.ts (hunk 5 above)
Comment (themadcreator) [reply packages/table/src/common/internal/scrollUtils.ts]: yeeesh. how many args would passing those props around save? this is an internal method and those args are passed by reference, so no perf penalty.

File: packages/table/src/common/internal/scrollUtils.ts [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+    let nextTop = currScrollTop;
+    let nextLeft = currScrollLeft;
+
+    // if these were max-frozen-index values, we would have added 1 before passing to the get*Offset
+    // functions, but the counts are already 1-indexed, so we can just pass those.
+    const frozenColumnsCumulativeWidth = getLeftOffset(numFrozenColumns);
+    const frozenRowsCumulativeHeight = getTopOffset(numFrozenRows);
+
+    if (cardinality === RegionCardinality.CELLS) {
```
Comment (themadcreator) [inline packages/table/src/common/internal/scrollUtils.ts]: for enums, i think we should use switch statements to make `default:` lintable.

File: packages/table/src/common/internal/scrollUtils.ts (hunk 11 above)
Comment (themadcreator) [reply packages/table/src/common/internal/scrollUtils.ts]: you can always put your switch-case in a block

```
case 1: {
   const foo = 1;
   break;
}
case 2: {
   const foo = 2;
   break;
}

Comment (themadcreator) [reply packages/table/src/common/internal/scrollUtils.ts]: `break`s are annoying yes, but the benefit of compile-time warnings shouldn't be discarded

File: packages/table/src/common/internal/scrollUtils.ts [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+        nextLeft = getAdjustedScrollPosition(leftOffset, frozenColumnsCumulativeWidth);
+    } else {
+        // if it's a FULL_TABLE region, scroll back to the top-left cell of the table
+        nextTop = 0;
+        nextLeft = 0;
+    }
+
+    return {
+        scrollLeft: nextLeft,
+        scrollTop: nextTop,
```
Comment (themadcreator) [inline packages/table/src/common/internal/scrollUtils.ts]: nit: could just call these scrollLeft/scrollRight then `return { scrollLeft, scrollRight }`

File: packages/table/src/common/internal/scrollUtils.ts [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    };
+}
+
+/**
+ * Adjust the scroll position to align content just beyond the frozen region, if necessary.
+ */
+function getAdjustedScrollPosition(scrollOffset: number, frozenRegionCumulativeSize: number) {
```
Comment (themadcreator) [inline packages/table/src/common/internal/scrollUtils.ts]: i appreciate the attempt at a meaningful name, but if i see "adjusted", i don't know what it's doing. maybe `getClampedScrollPosition`?

File: packages/table/src/quadrants/tableQuadrantStack.tsx [hunk 20]
```diff
@@ -192,6 +192,17 @@ export class TableQuadrantStack extends AbstractComponent<ITableQuadrantStackPro
         this.throttledHandleWheel = CoreUtils.throttleReactEventCallback(this.handleWheel, { preventDefault: true });
     }
 
+    /**
+     * Scroll the main quadrant to the specified scroll offset, keeping all other quadrants in sync.
+     */
+    public scrollToPosition(scrollLeft: number, scrollTop: number) {
+        const { scrollContainer } = this.quadrantRefs[QuadrantType.MAIN];
+        this.wasMainQuadrantScrollChangedFromOtherOnWheelCallback = false;
```
Comment (themadcreator) [inline packages/table/src/quadrants/tableQuadrantStack.tsx]: now THAT is a variable name. 💯

File: packages/table/src/table.tsx [hunk 21]
```diff
@@ -382,10 +383,13 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
     private refHandlers = {
         columnHeader: (ref: HTMLElement) => this.columnHeaderElement = ref,
         mainQuadrant: (ref: HTMLElement) => this.mainQuadrantElement = ref,
+        quadrantStack: (ref: TableQuadrantStack) => this.quadrantStackInstance = ref,
         rowHeader: (ref: HTMLElement) => this.rowHeaderElement = ref,
         scrollContainer: (ref: HTMLElement) => this.scrollContainerElement = ref,
     };
 
+    private quadrantStackInstance: TableQuadrantStack;
+
```
Comment (themadcreator) [inline packages/table/src/table.tsx]: nit: no need for a space here

File: packages/table/src/table.tsx [hunk 22]
```diff
@@ -435,6 +439,74 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
         };
     }
 
+    // Instance methods
```
Comment (themadcreator) [inline packages/table/src/table.tsx]: nit: these sort of block comments indicate the file is too big and should be split. "nit" because i know that splitting is extremely difficult at this point.

File: packages/table/src/table.tsx [hunk 23]
```diff
@@ -435,6 +439,74 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
+     * - FULL_TABLE: Scroll the top-left cell in the table to the top-left corner of the viewport.
+     *
+     * If there are active frozen rows and/or columns, the target region will be positioned in the
+     * top-left corner of the non-frozen area (unless the target region itself is in the frozen
+     * area).
+     *
+     * If the target region is close to the bottom-right corner of the table, this function will
+     * simply scroll the target region as close to the top-left as possible until the bottom-right
+     * corner is reached.
+     */
```
Comment (themadcreator) [inline packages/table/src/table.tsx]: i wish there were some way to annotate these methods so we could easily use the jsdoc comment in our docs. appreciate your update the docs markdown though

File: packages/table/src/table.tsx [hunk 24]
```diff
@@ -435,6 +439,74 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
+    public scrollToRegion(region: IRegion) {
+        const { left: currScrollLeft, top: currScrollTop } = this.state.viewportRect;
```
Comment (themadcreator) [inline packages/table/src/table.tsx]: holy crap, i didn't know you could rename with destructuring assignment!

File: packages/table/src/table.tsx [hunk 25]
```diff
@@ -435,6 +439,74 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
+
+        const { scrollLeft, scrollTop } = ScrollUtils.getScrollPositionForRegion(
+            region,
+            currScrollLeft,
+            currScrollTop,
+            this.grid.getCumulativeWidthBefore,
+            this.grid.getCumulativeHeightBefore,
+            numFrozenRows,
+            numFrozenColumns,
+        );
```
Comment (themadcreator) [inline packages/table/src/table.tsx]: looks like passing the objects would only save a couple args =/

Comment (themadcreator) [reply packages/table/src/table.tsx]: still, i prefer many args over jamming them into a custom argument object

File: packages/table/src/table.tsx [hunk 26]
```diff
@@ -1635,12 +1669,16 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
         const didScrollLeftChange = nextScrollLeft !== viewportRect.left;
 
         if (didScrollTopChange || didScrollLeftChange) {
-            // we need to modify the body element explicitly for the viewport to shift
+            // we need to modify the scroll container explicitly for the viewport to shift. in so
+            // doing, we add the size of the header elements, which are not technically part of the
+            // "grid" concept (the grid only consists of body cells at present).
             if (didScrollTopChange) {
-                this.bodyElement.scrollTop = nextScrollTop;
+                const topCorrection = this.shouldDisableVerticalScroll() ? 0 : this.columnHeaderElement.clientHeight;
```
Comment (themadcreator) [inline packages/table/src/table.tsx]: i think you use "adjustment" and "correction" interchangeably. choose one

File: packages/table/preview/index.tsx (hunk 15 above)
Comment (themadcreator) [reply packages/table/preview/index.tsx]: i guess the resizeRows does this. still, is this the idiomatic blueprint way to invoke actions on react components?

File: packages/table/src/common/internal/scrollUtils.ts (hunk 4 above)
Comment (cmslewis) [reply packages/table/src/common/internal/scrollUtils.ts]: I'd be into that. 👍

File: packages/table/src/common/internal/scrollUtils.ts (hunk 5 above)
Comment (cmslewis) [reply packages/table/src/common/internal/scrollUtils.ts]: It would save two args (`get{Top,Left}Offset => grid`, `currScroll{Top,Left} => viewportRect`), but the current args are much easier to mock for unit tests.

Also, we don't really _need_ to pass in the current scroll offsets. We could simply have this function return `null` for `scrollLeft` or `scrollTop` if that respective coordinate doesn't need to change, then the caller can substitute the current value for `null` in its scope. That reduces the params to:

```tsx
region: IRegion;
getLeftOffset: (columnIndex: number) => number;
getTopOffset: (rowIndex: number) => number;
numFrozenRows: number = 0;
numFrozenColumns: number = 0;
```

File: packages/table/src/common/internal/scrollUtils.ts (hunk 11 above)
Comment (cmslewis) [reply packages/table/src/common/internal/scrollUtils.ts]: @themadcreator nice, will change

File: packages/table/preview/index.tsx (hunk 15 above)
Comment (cmslewis) [reply packages/table/preview/index.tsx]: Yeah, there are a few instance methods around the codebase, each of which operates on a `ref`. The goal here was to provide a moment to briefly control the scroll without forcing the caller to control it forever and always (would be annoying and unnecessary). See:

```tsx
// Table
// =====

resizeRowsByTallestCell(columnIndices?: number | number[]): void;

// Toast
// =====

clear(): void;
dismiss(key: string): void;
getToasts(): IToastProps[];
show(props: IToastProps): string;
update(key: string, props: IToastProps): void;

// Tree
// ====

getNodeContentElement(nodeId: string | number): HTMLElement | undefined;
```

File: packages/table/src/common/internal/scrollUtils.ts (hunk 17 above)
Comment (cmslewis) [reply packages/table/src/common/internal/scrollUtils.ts]: Ah, good thought. Will change.

File: packages/table/src/table.tsx (hunk 21 above)
Comment (cmslewis) [reply packages/table/src/table.tsx]: K

File: packages/table/src/table.tsx (hunk 22 above)
Comment (cmslewis) [reply packages/table/src/table.tsx]: Agree. I want to change this later.

File: packages/table/src/table.tsx (hunk 23 above)
Comment (cmslewis) [reply packages/table/src/table.tsx]: Samesies.

File: packages/table/src/table.tsx (hunk 24 above)
Comment (cmslewis) [reply packages/table/src/table.tsx]: Yeah I just learned this last month. Super nifty. 👍

**Comment by cmslewis** [comment]
@giladgray @themadcreator think I addressed all of your immediate comments.
